File: embedded/esp8266/ino/lib/state_machine/state_machine.cpp
#include "state_machine.h"

#define DEFAULT_TIMEOUT_MS 1000

StateMachine::StateMachine(const uint8_t output_pin, uint32_t active_timeout_ms, uint32_t disabled_timeout_ms)
{
    _active_timeout = active_timeout_ms;
    _disabled_timeout = disabled_timeout_ms;

    _init_gpio(output_pin);
}

StateMachine::StateMachine(const uint8_t output_pin, uint32_t active_timeout_ms)
{
    _active_timeout = active_timeout_ms;
    _disabled_timeout = DEFAULT_TIMEOUT_MS;

    _init_gpio(output_pin);
}

StateMachine::StateMachine(const uint8_t output_pin)
{
    _active_timeout = DEFAULT_TIMEOUT_MS;
    _disabled_timeout = DEFAULT_TIMEOUT_MS;

    _init_gpio(output_pin);
}

void StateMachine::set_flag(void)
{
    _action_flag = true;
}

void StateMachine::run(void)
{
    switch (_state)
    {
        case STATE_IDLE: _on_state_idle(); break;
        case STATE_ACTIVE: _on_state_active(); break;
        case STATE_DISABLED: _on_state_disabled(); break;
        case STATE_ERROR: _on_state_error(); break;
        default: _set_state_error(); break;
    }
}

void StateMachine::_init_gpio(const uint8_t output_pin)
{
    _output_pin = output_pin;

    pinMode(output_pin, OUTPUT);
    digitalWrite(output_pin, LOW);
}

void StateMachine::_set_state_idle(void)
{
    _state = STATE_IDLE;
    _action_flag = false;
}

void StateMachine::_set_state_active(void)
{
    _state = STATE_ACTIVE;
    digitalWrite(_output_pin, HIGH);

    _stopwatch = millis();
}

void StateMachine::_set_state_disabled(void)
{
    _state = STATE_DISABLED;
    digitalWrite(_output_pin, LOW);

    _stopwatch = millis();
}

void StateMachine::_set_state_error(void)
{
    _state = STATE_ERROR;
    digitalWrite(_output_pin, LOW);
    _action_flag = false;
}

void StateMachine::_on_state_idle(void)
{
    if (_action_flag)
    {
        _set_state_active();
    }
}

void StateMachine::_on_state_active(void)
{
    uint32_t dt = millis() - _stopwatch;
    if (dt > _active_timeout)
    {
        _set_state_disabled();
    }
}

void StateMachine::_on_state_disabled(void)
{
    uint32_t dt = millis() - _stopwatch;
    if (dt > _disabled_timeout)
    {
        _set_state_idle();
    }
}

void StateMachine::_on_state_error(void)
{
}

File: embedded/python/apps/virtual_device/relay_controller/state_machine.py
import time

from libs.io.digital_io_controller import DigitalIOController

STATE_IDLE = 0
STATE_ACTIVE = 1
STATE_DISABLED = 2
STATE_ERROR = -1

DEFAULT_TIMEOUT_MS = 1000

class StateMachine:
    __state = STATE_IDLE
    __action_flag = False
    __stopwatch = 0

    def __init__(self, output: DigitalIOController, active_timeout_ms: int = DEFAULT_TIMEOUT_MS, disabled_timeout_ms: int = DEFAULT_TIMEOUT_MS):
        self.__active_timeout = active_timeout_ms * 0.001
        self.__disabled_timeout = disabled_timeout_ms * 0.001
        self.__output_controller = output

        self.__set_state_idle()

    def set_flag(self):
        self.__action_flag = True

    def run(self):
        if self.__state == STATE_IDLE:
            self.__on_state_idle()
        elif self.__state == STATE_ACTIVE:
            self.__on_state_active()
        elif self.__state == STATE_DISABLED:
            self.__on_state_disabled()
        elif self.__state == STATE_ERROR:
            self.__on_state_error()
        
        else:
            self.__set_state_error()

    def __set_state_idle(self):
        self.__state = STATE_IDLE
        self.__action_flag = False

    def __set_state_active(self):
        self.__state = STATE_ACTIVE
        self.__output_controller.on()

        self.__stopwatch = time.time()

    def __set_state_disabled(self):
        self.__state = STATE_DISABLED
        self.__output_controller.off()
        
        self.__stopwatch = time.time()

    def __set_state_error(self):
        self.__state = STATE_ERROR
        self.__output_controller.off()
        self.__action_flag = False

    def __on_state_idle(self):
        if self.__action_flag:
            self.__set_state_active()

    def __on_state_active(self):
        dt = time.time() - self.__stopwatch
        if dt > self.__active_timeout:
            self.__set_state_disabled()

    def __on_state_disabled(self):
        dt = time.time() - self.__stopwatch
        if dt > self.__disabled_timeout:
            self.__set_state_idle()

    def __on_state_error(self):
        pass
